odoo/tools/sql.py

```diff
@@ -122,7 +121,7 @@ def __init__(self, code: (str | SQL) = "", /, *args, to_flush: (Field | None) =
         if to_flush is not None:
             to_flush_list.append(to_flush)
 
-        self.__code = code % tuple(code_list)
```

Comment:
The caller should escape properly the arguments and this constructor just checks them.

Reply:
We perform double formatting when we have a list of arguments. One inside the `__init__` then another by [`cr.execute`](https://github.com/odoo/odoo/blob/3cae03ff9e1f5f25993228718b5fb465db6a0442/odoo/sql_db.py#L343-L354).

From the diff discussed before:
```diff
iff --git a/odoo/tools/sql.py b/odoo/tools/sql.py
index bd77e6825b50..4e498202a00d 100644
--- a/odoo/tools/sql.py
+++ b/odoo/tools/sql.py
@@ -102,7 +102,7 @@ class SQL:
         if kwargs:
             code, args = named_to_positional_printf(code, kwargs)
         elif not args:
-            code % ()  # check that code does not contain %s
+            code = code % ()  # check that code does not contain %s
             self.__code = code
             self.__params = ()
             self.__to_flush = () if to_flush is None else (to_flush,)
@@ -440,11 +440,9 @@ def constraint_definition(cr, tablename, constraintname):
 
 def add_constraint(cr, tablename, constraintname, definition):
     """ Add a constraint on the given table. """
-    if "%" in definition:
-        definition = definition.replace("%", "%%")
     query1 = SQL(
         "ALTER TABLE %s ADD CONSTRAINT %s %s",
-        SQL.identifier(tablename), SQL.identifier(constraintname), SQL(definition),
+        SQL.identifier(tablename), SQL.identifier(constraintname), SQL(definition.replace("%", "%%")),
     )
     query2 = SQL(
         "COMMENT ON CONSTRAINT %s ON %s IS %s",
```

The problem is that there is code out there assuming that we do not perform a formatting if there is no args.
https://github.com/odoo/enterprise/blob/0f7a4eb991fc59240710c0f8ba78f4da5b889fe6/l10n_cl_reports/models/account_eightcolumns_report.py#L117
See also: https://runbot.odoo.com/runbot/build/71220872

It causes errors at execution time if we mix SQL arguments with other type of arguments.
```py
>>> cr.execute(sql.SQL("SELECT 1 %s AND 'x'=%s", sql.SQL("WHERE 'ab' LIKE 'a%%'"), "x"))  # fails
...
IndexError: list index out of range
```
The passed `WHERE` gets formatted.

Thus the point of the alternative patch above is to try to not break existing code :/ I'm not happy the it either. Suggestions are welcome.

Note that I believe the check `code % ()` is a bit odd. For example this is fine `sql.SQL("SELECT 1 WHERE 'ab' LIKE 'a%'")` and will be executed by psycopg without issues (it doesn't perform any formatting if there are no arguments).

Reply:
The invariant in this class is that `self.__code % self.__params` should always work.
Removing the check `code % ()` results in the following being accepted `SQL("%s %s", SQL("test%s"), 3)` and invalid.

Otherwise, you need to update also the composition of SQL's to both handle the cases that either have or don't have arguments. That would be an ugly solution until we fix existing code.

Reply:
> The invariant in this class is that `self.__code % self.__params` should always work.

I think the invariant is broken: there is no need to format if there are no params. This is consistent with psycopg behaviour. We are tied to that since we later pass the `code` and `params` to psycog `execute`.

> Removing the check code % () results in the following being accepted `SQL("%s %s", SQL("test%s"), 3)` and invalid.

OK, but this is valid `SQL("7%5")` and fails current check. We perhaps prefer to err on the safe side... but how would you pass an expression that uses the `%` operator with current check? This is, of course, just an example.

> Otherwise, you need to update also the composition of SQL's to both handle the cases that either have or don't have arguments. That would be an ugly solution until we fix existing code.

Indeed, that's what I fear...

Reply:
My suggestion: keep `__code` always escaped, in case we have no arguments, accept the code by doubling "%".

```python
if kwargs:
    code, args = named_to_positional_printf(code, kwargs)  # TODO adapt function to double the "%%"
elif not args:
    if '%s' not in code:
        # double "%" so that they are escaped
        # skip strings with %s which should be properly escaped
        code = code.replace('%', '%%')
    code % ()  # check that code can be interpolated
    self.__code = code
    self.__params = ()
    self.__to_flush = () if to_flush is None else (to_flush,)
    return

code = code.replace('%%', '%%%%')  # double "%%" because we interpolate the string and we want to keep it
```


@ryv-odoo What do you think?

Reply:
I've pushed a suggestion to move on with the constraint's issue. I think further patches need to be done by you guys... you have much more context and details to test all this correctly ;) Feel free to push here.

There are still some caveats IMO with your suggestion above. Imagine we want to do add some sort of "no-plurals" filtering: `noun NOT LIKE '%s'` :/ In other words the root of the issue is that `%` as well as `%s` are valid tokens without any formatting meaning. Ensuring code out there still works is difficult. :/